machine-learning/skorch-hyperparam-opt.ipynb

```diff
@@ -0,0 +1,1117 @@
+    "                batch_size=32,\n",
+    "                verbose=2)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "  epoch    train_loss    valid_acc    valid_loss      dur\n",
+      "-------  ------------  -----------  ------------  -------\n",
+      "      1        \u001b[36m0.6302\u001b[0m       \u001b[32m0.7516\u001b[0m        \u001b[35m0.5243\u001b[0m  66.5468\n",
+      "      2        \u001b[36m0.4797\u001b[0m       \u001b[32m0.7994\u001b[0m        \u001b[35m0.4413\u001b[0m  59.9177\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "<class 'skorch.classifier.NeuralNetClassifier'>[initialized](\n",
+       "  module_=CNN(\n",
+       "    (embedding): Embedding(400000, 100)\n",
+       "    (conv_0): Conv1d(1, 100, kernel_size=(2, 100), stride=(1,))\n",
+       "    (conv_1): Conv1d(1, 100, kernel_size=(3, 100), stride=(1,))\n",
+       "    (conv_2): Conv1d(1, 100, kernel_size=(4, 100), stride=(1,))\n",
+       "    (fc): Linear(in_features=300, out_features=2, bias=True)\n",
+       "    (dropout): Dropout(p=0.2, inplace=False)\n",
+       "  ),\n",
+       ")"
+      ]
+     },
+     "execution_count": 23,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "skorch_model.fit(X_train, y_train)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 24,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "0.79112"
+      ]
+     },
+     "execution_count": 24,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# quick check on the test set accuracy\n",
+    "skorch_model.score(X_test, y_test)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 25,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "(array([0, 1]), array([12500, 12500]))"
+      ]
+     },
+     "execution_count": 25,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# random guessing would 50% accuracy so the model is indeed training\n",
+    "np.unique(y_test, return_counts=True)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "del skorch_model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Grid search with Hyperband"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# https://ml.dask.org/hyper-parameter-search.html#hyperband-parameters-rule-of-thumb\n",
+    "EPOCHS = 5\n",
+    "NUM_TRAINING_EXAMPLES = len(train)*.8\n",
+    "n_examples = EPOCHS * NUM_TRAINING_EXAMPLES\n",
+    "n_params = 8\n",
```

Comment:
I think this should be added the docstring too.

Reply:
I think this is a really helpful comparison for people - to see how you're able to sample many more parameters with Hyperband.

I can increase `n_params` but my frame of reference was that `n_params` is functionally equivalent to `n_iter` in RandomizedSearchCV, where higher numbers just mean more compute. If `n_params` can be increased without increasing the compute cost, then fantastic! If that's the case, then _this_ would be useful to include in the docs.

I need to understand how more params impact GPU memory utilization as well. If models are only initialized one at a time (assuming 1 single GPU) then this is likely fine. The key thing is that those models need to be unpersisted somehow, otherwise you will fill up GPU memory.